packages/leancode_flutter_svg_adaptive_loader/lib/src/leancode_flutter_svg_adaptive_loader.dart

```diff
@@ -0,0 +1,149 @@
+import 'dart:convert';
+import 'dart:math';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_svg/flutter_svg.dart';
+import 'package:vector_graphics/vector_graphics.dart';
+import 'package:vector_graphics_compiler/vector_graphics_compiler.dart' as vg;
+
+/// A [BytesLoader] able to distinguish xml-based svg from compiled binary form,
+/// by looking for '<svg' pattern at 1000 of first characters.
+/// Using logic from [AssetBytesLoader] for binaries and [SvgAssetLoader]
+/// for standard xml-based format.
+class LeancodeFlutterSvgAdaptiveLoader extends BytesLoader {
+  /// Creates a [BytesLoader]
+  const LeancodeFlutterSvgAdaptiveLoader(
+    this.assetName, {
+    this.packageName,
+    this.assetBundle,
+  });
+
+  /// The name of the asset to load.
+  final String assetName;
+
+  /// The package name to load from, if any.
+  final String? packageName;
+
+  /// The asset bundle to use.
+  ///
+  /// If unspecified, [DefaultAssetBundle.of] the current context will be used.
+  final AssetBundle? assetBundle;
+
+  AssetBundle _resolveBundle(BuildContext? context) {
+    if (assetBundle != null) {
+      return assetBundle!;
+    }
+    if (context != null) {
+      return DefaultAssetBundle.of(context);
+    }
+    return rootBundle;
+  }
+
+  Future<ByteData> _prepareMessage(BuildContext? context) {
+    return _resolveBundle(context).load(
+      packageName == null ? assetName : 'packages/$packageName/$assetName',
+    );
+  }
+
+  String _decode(ByteData? message, [int? lenght]) =>
+      utf8.decode(message!.buffer.asUint8List(0, lenght), allowMalformed: true);
+
+  /// This method intentionally avoids using `await` to avoid unnecessary event
+  /// loop turns. This is meant to to help tests in particular.
+  @override
+  Future<ByteData> loadBytes(BuildContext? context) {
+    return _prepareMessage(context).then((message) {
+      // Dart encodes strings in UTF-16 so one codepoint is 2 bytes
+      final first1000Chars = _decode(message, min(2000, message.lengthInBytes));
+
+      if (first1000Chars.contains('<svg')) {
```

Comment:
Has this been studied to be reliable? It looks very hacky. I'll give it a shot to see if I can generate an example that breaks this check

Reply:
Follow up: check https://github.com/dnfield/vector_graphics/issues/210. Author is not keen on making the magic number public. But some other solutions were proposed instead (IMO all less than ideal ones). See if we can use any of them.

Reply:
Im not sure if our current implementation (looking for "<svg") requires any change? Do you think I should replace "<svg" with "<" alone?

Reply:
If we are going the `<` route it would be better if we lazily decoded utf8 to look for it, instead of arbitrarily getting 2000 chars. The 2000 chars can still be broken we we have a lot of leading whitespace. I can not find a way to lazily decode bytes to utf8 though. Maybe @Albert221 you have some ideas here?

Reply:
I believe it is a limitation that we can agree on, cuz it is very unlikely for some asset to have 1000 chars of white space / comments (only case I see is some kind of documentation for the vector)

Reply:
I'm thinking of a similar check: check first _n_ bytes for whether they're UTF8-encoded. If they are - treat it as SVG. If they're not - they're most probably the binary vector format

Reply:
Yup, let's do this:

Read first _n_ bytes (let's make it 100), with utf8 decoder with `allowMalformed: true` and count `\uFFFD` characters (�; this is a character that the decoder places in place of invalid codepoints) in the code. If it's more than 1, we got invalid UTF-8 encoded string - we treat it as binary vector. If it's 1, we clipped the last codepoint when we read n bytes. It's fine. We treat it as SVG. If it's 0 - same story - we treat it as SVG.